aten/src/ATen/templates/TensorBody.h

```diff
@@ -801,22 +806,25 @@ class TORCH_API Tensor {
 #pragma warning( pop )
 #endif
 
-int64_t get_device(Tensor self);
+inline int64_t get_device(Tensor self) {
+  return self.get_device();
+}
 
 template <typename T>
 auto Tensor::register_hook(T&& hook) const -> Tensor::hook_return_void_t<T> {
   // Return the grad argument in case of a hook with void return type to have an
   // std::function with Tensor return type
-  std::function<void(Tensor)> fn(hook);
-  return _register_hook([fn](const Tensor& grad) {
+  static_assert(std::is_same<decltype(hook(Tensor())), void>::value,
+                "Expected hook to return void");
+  return _register_hook([fn=std::forward<T>(hook)](const Tensor& grad) {
```

Comment:
Somewhat unrelated other than being in the same file. This prevents nesting two levels of `std::function`.

Reply:
Please split out unrelated changes.